src/adapters/WstEthChainlinkAdapter.sol

```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity 0.8.21;
+
+import {IStEth} from "../interfaces/IStEth.sol";
+import {AggregatorV3Interface} from "../interfaces/AggregatorV3Interface.sol";
+
+import {ErrorsLib} from "../libraries/ErrorsLib.sol";
+
+/// @title WstEthChainlinkAdapter
+/// @author Morpho Labs
+/// @custom:contact security@morpho.org
+/// @notice wstETH/ETH exchange rate price feed.
+/// @dev This contract should only be used as price feed for `ChainlinkOracle`.
+contract WstEthChainlinkAdapter is AggregatorV3Interface {
+    uint8 public constant decimals = uint8(18);
+    string public constant description = "wstETH/ETH exchange rate";
+
+    IStEth public immutable ST_ETH;
+
+    constructor(address stEth) {
+        require(stEth != address(0), ErrorsLib.ZERO_ADDRESS);
+        ST_ETH = IStEth(stEth);
+    }
+
+    /// @notice Reverts as no Chainlink aggregator is used.
+    function version() external pure returns (uint256) {
+        revert();
+    }
+
+    /// @notice Reverts as it's not necessary for the ChainlinkOracle contract.
+    function getRoundData(uint80) external pure returns (uint80, int256, uint256, uint256, uint80) {
+        revert();
+    }
+
+    function latestRoundData() external view returns (uint80, int256, uint256, uint256, uint80) {
+        uint256 answer = ST_ETH.getPooledEthByShares(10 ** decimals);
```

Comment:
I think that using decimals here is bad, because it would well be that pricing with 10^decimals is not precise enough. On Ethereum totalShares ~ totalETH, so it's safe but it could well not the case on an other chain. Imagine the following case: you have totalShares = 1e36, and you fork this contract to have only 8 decimals (why not, half chainlink feeds are like that). The price of one wstETH in stETH (=ETH) is well defined (let's say 1), but this oracle will return zero.

A design with a conversionSample would be clearer for me and safer (even if it would do almost exactly the same computations).

We can also use `stETHperToken` from wstETH:
https://github.com/lidofinance/lido-dao/blob/331ecec7fe3c8d57841fd73ccca7fb1cc9bc174e/contracts/0.6.12/WstETH.sol#L107-L109
(I know that it performs exactly the same computation in the end, but it's safer)

```suggestion
        uint256 answer = WSTETH.stETHPerToken() * 10 ** (WSTETH.decimals() - decimals);
```

Reply:
up @MathisGD

Reply:
I agree we should keep the code specifically crafted for wstETH on mainnet, but add a comment about this

Reply:
>why is it safer if we perform exactly the same thing?

it's explained in my first comment no? I think that we confound a little bit the oracle precision, the conversion sample and the wstETH decimals and it's clearer with my version

Reply:
Here, we're adding a call while we know that this is tailored to wstETH which is immutable. I proposed to add a comment instead, isn't it sufficient? If it's a blocker for you I would use a constant instead

Reply:
we don't need to add a call, we can take store the result as an immutable

Reply:
> we don't need to add a call, we can take store the result as an immutable

This is best of both world

Reply:
@MathisGD isn't it more something like: `WSTETH.stETHPerToken() * 10 ** decimals / 10 ** WSTETH.decimals();`?

Reply:
Or we can more simply hardcode: `decimals = 18` and `return WST_ETH.stEthPerToken()` with a comment and that's also fine

If I'm not mistaken, the issue is we use `decimals` to quote the price while they have 2 distinct roles. So if we're hardcoding decimals and hardcoding the sample, independent from the decimals that should be fine

Reply:
I agree with @Rubilmax solution, where we can hardcode the decimals since this oracle is not supposed to be used on other chains (or is it ?)

Reply:
It can be used on another chain but the `wstETH` implementation is known in advance so to me this solution is fine!